python/cudf_polars/cudf_polars/utils/config.py

```diff
@@ -5,209 +5,306 @@
 
 from __future__ import annotations
 
-import copy
+import dataclasses
+import enum
 import json
-from typing import TYPE_CHECKING, Any
+from typing import TYPE_CHECKING, Literal
 
 if TYPE_CHECKING:
-    from collections.abc import MutableMapping, Sequence
-
     from typing_extensions import Self
 
+    import polars as pl
+
+
 __all__ = ["ConfigOptions"]
 
 
-class ConfigOptions:
+# TODO: Use enum.StrEnum when we drop Python 3.10
+
+
+class FallbackMode(str, enum.Enum):
     """
-    GPUEngine configuration-option manager.
+    How to handle errors when the GPU engine fails to execute a query.
 
-    This is a convenience class to help manage the nested
-    dictionary of user-accessible `GPUEngine` options.
+    * ``FallbackMode.WARN`` : Emit a warning and fall back to the CPU engine.
+    * ``FallbackMode.RAISE`` : Raise an exception.
+    * ``FallbackMode.SILENT``: Silently fall back to the CPU engine.
     """
 
-    __slots__ = ("_hash_value", "config_options")
-    _hash_value: int
-    config_options: dict[str, Any]
-    """The underlying (nested) config-option dictionary."""
+    WARN = "warn"
+    RAISE = "raise"
+    SILENT = "silent"
 
-    def __init__(self, options: dict[str, Any]):
-        self.config_options = self.validate(copy.deepcopy(options))
 
-    def set(self, name: str, value: Any) -> Self:
-        """
-        Set a user config option.
+class Scheduler(str, enum.Enum):
+    """
+    The scheduler to use for the streaming executor.
+
+    * ``Scheduler.SYNCHRONOUS`` : Use the synchronous scheduler.
+    * ``Scheduler.DISTRIBUTED`` : Use the distributed scheduler.
+    """
 
-        Nested dictionary keys should be separated by periods.
-        For example::
+    SYNCHRONOUS = "synchronous"
+    DISTRIBUTED = "distributed"
 
-            >>> options = options.set("parquet_options.chunked", False)
 
-        Parameters
-        ----------
-        name
-            Period-separated config name.
-        value
-            New config value.
-        """
-        options = config_options = copy.deepcopy(self.config_options)
-        keys = name.split(".")
-        for k in keys[:-1]:
-            assert isinstance(options, dict)
-            if k not in options:
-                options[k] = {}
-            options = options[k]
-        options[keys[-1]] = value
-        return type(self)(config_options)
-
-    def get(self, name: str, *, default: Any = None) -> Any:
-        """
-        Get a user config option.
+class ShuffleMethod(str, enum.Enum):
+    """
+    The method to use for shuffling data between workers.
 
-        Nested dictionary keys should be separated by periods.
-        For example::
+    * ``ShuffleMethod.TASKS`` : Use the task-based shuffler.
+    * ``ShuffleMethod.RAPIDSMFP`` : Use the rapidsmpf scheduler.
+    """
 
-            >>> chunked = config_options.get("parquet_options.chunked")
+    TASKS = "tasks"
+    RAPIDSMFP = "rapidsmpf"
```

Comment:
We probably need an "auto" option to emulate the `None` default we had previously (i.e. we should try using "rapidsmp" when we have a distributed schedule and rapidsmpf is available, otherwise use `"tasks"`)

Reply:
I'll document this, but on `ConfigOptions` this is `shuffle_method: ShuffleMethod | None` and the default is to try rapidsmpf and fall back.

We *could* add `"auto"` as an allowed value here, but that would

Reply:
>I'll document this, but on ConfigOptions this is shuffle_method: ShuffleMethod | None and the default is to try rapidsmpf and fall back.

Ah, okay.

>We could add "auto" as an allowed value here, but that would maybe require importing rapidsmp when we initialize the config.

I was imagining the case that `"auto"` **is** the shuffle method as far as the `ShuffleMethod` is concerned. We wouldn't check if rapidsmpf was available until we needed to build the graph.

Reply:
Cool, that's what we should have now, it's just via `executor.shuffle_method is None`.